### PR TITLE
raft: avoid copy prs

### PR DESCRIPTION
**Pull request by hicqu**
We can use `Option<FlatMap<...>>` instead of `FlatMap` in order to avoid copy that.

### CODE REVIEW

File: src/raft/raft.rs [hunk 1]
```diff
@@ -566,16 +570,14 @@ impl<T: Storage> Raft<T> {
...
-            if id == self.id {
-                continue;
-            }
-            self.send_append(id);
-        }
+        let self_id = self.id;
+        let mut prs = self.take_prs();
+        prs.iter_mut()
+            .filter(|&(id, _)| *id != self_id)
+            .for_each(|(id, ref mut pr)| { self.send_append(*id, pr); });
```
Comment (BusyJay) [inline src/raft/raft.rs]: Brackets can be removed.

Comment (BusyJay) [reply src/raft/raft.rs]: Is `ref mut` necessary?

File: src/raft/raft.rs [hunk 2]
```diff
@@ -648,10 +648,11 @@ impl<T: Storage> Raft<T> {
             e.set_index(li + 1 + i as u64);
         }
         self.raft_log.append(es);
-        self.prs
-            .get_mut(&self.id)
+        let mut prs = self.take_prs();
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why take?

File: src/raft/raft.rs [hunk 3]
```diff
@@ -1219,66 +1266,28 @@ impl<T: Storage> Raft<T> {
         maybe_commit: &mut bool,
         more_to_send: &mut Option<Message>,
     ) {
-        if !self.prs.contains_key(&m.get_from()) {
+        if self.get_prs().get(&m.get_from()).is_none() {
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why not use `contains`?

File: src/raft/raft.rs [hunk 4]
```diff
@@ -1431,7 +1442,10 @@ impl<T: Storage> Raft<T> {
         }
 
         if send_append {
-            self.send_append(m.get_from());
+            let from = m.get_from();
+            let mut prs = self.take_prs();
```
Comment (BusyJay) [inline src/raft/raft.rs]: Is it really better than the original implementations?

Comment (hicqu) [reply src/raft/raft.rs]: Here it's not better. But `send_append` called in `handle_transfer_leader` needs a `pr` as argument.

File: src/raft/raft.rs (hunk 3 above)
Comment (hicqu) [reply src/raft/raft.rs]: `FlatMap` doesn't have method `contains`.

File: src/raft/raft.rs [hunk 5]
```diff
@@ -528,15 +528,13 @@ impl<T: Storage> Raft<T> {
...
-            }
-            (
-                self.raft_log.term(pr.next_idx - 1),
-                self.raft_log.entries(pr.next_idx, self.max_msg_size),
-            )
-        };
+        if pr.is_paused() {
+            return;
+        }
+        let (term, ents) = (
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why put them into tuple?

File: tests/raft/test_raft.rs [hunk 6]
```diff
@@ -2539,8 +2559,8 @@ fn test_bcast_beat() {
     let mut msgs = sm.read_messages();
     assert_eq!(msgs.len(), 2);
     let mut want_commit_map = HashMap::new();
-    want_commit_map.insert(2, cmp::min(sm.raft_log.committed, sm.prs[&2].matched));
-    want_commit_map.insert(3, cmp::min(sm.raft_log.committed, sm.prs[&3].matched));
+    want_commit_map.insert(2, cmp::min(sm.raft_log.committed, sm.mut_prs()[&2].matched));
```
Comment (BusyJay) [inline tests/raft/test_raft.rs]: Why mut?

File: src/raft/raft.rs (hunk 3 above)
Comment (BusyJay) [reply src/raft/raft.rs]: How about `contains_key`?

File: src/raft/raft.rs (hunk 4 above)
Comment (BusyJay) [reply src/raft/raft.rs]: Can `send_append` accept an `Option`?

**Comment by arthurprs** [comment]
Another way is to replace the FlatMap with an empty one with `mem::swap` or `mem::replace`. I often find that to be easier on the eyes and also on the cpu.

File: src/raft/raft.rs (hunk 4 above)
Comment (hicqu) [reply src/raft/raft.rs]: It's really a good idea? Sometimes it gets `pr` from external arguments, but sometimes from `self.get_prs()`? I think make it unify is also importent for readability.

**Comment by siddontang** [comment]
LGTM

But we must be very careful to not to return early before resetting the progress, maybe this is a risk later.

**Comment by hicqu** [comment]
@arthurprs That a way I have tried. But because `prs` is still avaliable after `mem::replace`, it's very hard to write correct code.

**Comment by hicqu** [comment]
@BusyJay PTAL again, thanks.

File: src/raft/raft.rs [hunk 7]
```diff
@@ -801,7 +799,7 @@ impl<T: Storage> Raft<T> {
             }
             return;
         }
-        let ids: Vec<_> = self.prs.keys().cloned().collect();
+        let ids: Vec<_> = self.get_prs().keys().cloned().collect();
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why not `take_prs`?

File: src/raft/raft.rs [hunk 8]
```diff
@@ -566,16 +566,14 @@ impl<T: Storage> Raft<T> {
...
-            if id == self.id {
-                continue;
-            }
-            self.send_append(id);
-        }
+        let self_id = self.id;
+        let mut prs = self.take_prs();
+        prs.iter_mut()
+            .filter(|&(id, _)| *id != self_id)
+            .for_each(|(id, pr)| { self.send_append(*id, pr); });
```
Comment (BusyJay) [inline src/raft/raft.rs]: Can it keep consistent with L588~L590

File: src/raft/raft.rs [hunk 9]
```diff
@@ -585,14 +583,12 @@ impl<T: Storage> Raft<T> {
...
-            if id == self.id {
-                continue;
-            }
-            self.send_heartbeat(id, ctx.clone());
-        }
+        let self_id = self.id;
+        let mut prs = self.take_prs();
+        prs.iter_mut()
+            .filter(|&(id, _)| *id != self_id)
+            .for_each(|(id, pr)| { self.send_heartbeat(*id, pr, ctx.clone()); });
```
Comment (BusyJay) [inline src/raft/raft.rs]: Why put them into `{}`?

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**